lib/libcgroup-3.1.0/doc/man/cgdelete.1
.\" Copyright (C) 2009 Red Hat, Inc. All Rights Reserved.
.\" Written by Jan Safranek <jsafrane@redhat.com>

.TH CGDELETE  1 2009-10-26 "Linux" "libcgroup Manual"
.SH NAME

cgdelete \- remove control group(s)

.SH SYNOPSIS
\fBcgdelete\fR [\fB-h\fR] [\fB-r\fR] [\fB-b\fR] [[\fB-g\fR]
<\fIcontrollers\fR>:\fI<path\fR>] ...

.SH DESCRIPTION
The \fBcgdelete\fR
program removes all specified control groups.

.TP
.B -b
ignores the default systemd delegated hierarchy path and
constructs the path of the control groups relative to the
cgroup root hierarchy.

.TP
.B [-g] <controllers>:<path>
Defines the control group to delete. Multiple control groups may be
specified.
.B -g
is optional.

.TP
.B -h, --help
Display this help and exit.

.TP
.B -r, --recursive
Recursively remove all subgroups.

.SH ENVIRONMENT VARIABLES
.TP
.B CGROUP_LOGLEVEL
controls verbosity of the tool. Allowed values are \fBDEBUG\fR,
\fBINFO\fR, \fBWARNING\fR or \fBERROR\fR.

.SH EXAMPLES
.TP
.B cgdelete -g cpu,devices:/test
remove control group test from hierarchies containing cpu and device controllers


.SH SEE ALSO
cgcreate (1), lscgroup (1)


ChangeLog
DOMjudge Programming Contest Judging System

Version 8.4.0DEV
---------------------------

Version 8.3.0 - 31 May 2024
---------------------------
 - [security] Close metadata file descriptor for the child in runguard.
 - Document that minimum PHP version is now 8.1.0.
 - Upgrade Symfony to 6.2 and upgrade other library dependencies.
 - Show link to problem text on jury problem overview and contest detail pages.
 - Add label to teams to be in sync with ICPC CCS spec.
 - Add feedback/teammessage.txt for automated teamfeedback, see: #2016.
 - Clarifications can now use markdown syntax.
 - Upgrade Bootstrap to 5.2 and dependent components.
 - Bug fixed to correctly select current user during configuration of installation.
 - Properly fix removal of problem/submission data.
 - Filter away balloons of jury members in balloon runner interface.
 - Fix upgrade path for executables, see: #1826.
 - Collect compiler and runner versions before compilation, useful to compare
   judgehosts.
 - Various improvements to shadow config used at ICPC World Finals.
 - Show the historic timelimit which was used when judging the submission.
 - Fixes to the documentation and import of problems with @EXPECTED_RESULTS@
   tags.
 - Allow filtering on affilition in submission page for jury for multisite
   contests.
 - Add preview of the contest problemset to the jury page before contest
   start.
 - Add optional Adminer for easy database access.
 - Fix medal awards when skipping categories.
 - Add direct button to public page for registering when enabled.
 - Scale batch size of judge tasks dynamically.
 - Rename room to location for teams.
 - Add option to upload contest problemset document.
 - Improve on the advice for rejuding on config changes, only when relevant
   advice to start a rejudging, similar for updating the scoreboard cache.
 - For interactive problems: add explicit statement to when one of sides of
   the interactions ran closed their output.
 - Allow importing a (generic) contest warning which is displayed on the scoreboard.
 - Added option in the dj_setup_database to make dumps, useful before major
   upgrades and during the contest before problem imports.
 - Fixed some bugs to become OpenAPI spec compliant again.
 - Prevent a race condition leading to two submissions receiving the same
   colour balloon.
 - Compile executables with `-std=gnu++20` now.
 - Record the testcase directory for judging runs, this makes investigating
   easier.
 - Set default C/C++ standards in configure script, we intend to ship all
   scripts free of warnings/errors.
 - Respect basic authentication in the URL for the import-contest when .netrc
   exists.
 - Fix CORS header for API to make extending with other tools easier.
 - Fix eventfeed when SQL commits happen out of order to send all events.
 - Lazy load the team pictures for the scoreboard, to prevent spamming access
   log with HTTP404s
 - Make TLE detection more robust for interactive problems.
 - Added option for ICPC online scoreboard with no rank displays.

Version 8.2.0 - 6 March 2023
----------------------------
 - Various bugfixes & UI improvements.
 - Add support for Rust as submission language.
 - Add teamphotos to scoreboard zip.
 - Add jury option to to download ZIP with all problem samples, attachments
   and problem statement.
 - Extend team import with optional room and other API extensions, see the API
   doc for further details.
 - Drop support for shadowing a single contest from multiple sources.
 - Log account logins via API separately.
 - Add support for PHP v8.2.
 - Upgraded library dependencies.
 - Removed output of detected entry point to teams.
 - Added option to add custom apt repositories to judge chroot.
 - Added CLI script to configure the domserver with a yaml file.
 - Changed activation behaviour of contests imported via UI to be directly
   available.
 - New option added to reset login status of teams between contests.
 - Highlight submissions which were judged with compiler flags which since
   have changed. Useful when you find wrong flags during the contest.
 - Handle network errors for judgedaemon when domserver is unavailable.
 - Added super lazy mode useful for ICPC analyst instances. Here nothing is
   judged unless manually requested.
 - Split problem import messages in info/warning/danger, to make sure errors
   are not missed when importing.
 - Store user login sessions in database.
 - Contest import CLI script is now rewritten in python.
 - The example test problems are now imported via the contest package format.

Version 8.1.0 - 1 June 2022
---------------------------
 - Document that minimum PHP version is now 7.4.0.
 - Add support for PHP 8.1.
 - Upgrade Symfony to 5.4 LTS and upgrade other library dependencies.
 - Replace calls to deprecated `strftime` with `date`. Note: if you have
   configured a custom `time_format`, you need to update it to the format
   supported by `date`.
 - Dropped the presentation-error verdict, databases from before DOMjudge 5.0
   will be altered to replace the verdict with wrong-answer.
 - Changed API import to give admin users a team.
 - Balloon users can now be created via the API.
 - Shadow mode will now fix missing data by reading from the primary CCS.
 - Fixed cgroup naming for docker to prevent sharing the same container.
 - Removed python2 in the chroot, remove python3 in favor of pypy3 in chroot.
 - Added support for sentry for logging.
 - Added default external IDs on all places.
 - Fixed medals on scoreboard for filtered scoreboards.
 - Dropped old balloons CLI.
 - Reimplementation of runpipe for interactive problems.
 - Individual files can now be removed from executables.
 - Added mysql socket option, for when port/IP is not used.
 - Added support for accounts.{yaml,json} for import script.
 - Added support for judgeremaining for problems.
 - Add option to delete existing testcases/attachments/other data when
   importing a problem.

lib/libcgroup-3.1.0/src/tools/cgdelete.c
// SPDX-License-Identifier: LGPL-2.1-only
/**
 * Copyright RedHat Inc. 2009
 *
 * Authors:	Jan Safranek <jsafrane@redhat.com>
 */

#include "tools-common.h"

#include <libcgroup.h>
#include <libcgroup-internal.h>

#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <stdio.h>
#include <errno.h>

static const struct option  long_options[] = {
	{"recursive",	      no_argument, NULL, 'r'},
	{"help",	      no_argument, NULL, 'h'},
	{"group",	required_argument, NULL, 'g'},
	{NULL, 0, NULL, 0}
};

struct ext_cgroup_record {
	char name[FILENAME_MAX];	/* controller name */
	char controller[FILENAME_MAX];	/* cgroup name */
	int h_number;			/* hierarchy number */
};


static void usage(int status, const char *program_name)
{
	if (status != 0) {
		err("Wrong input parameters,");
		err(" try %s --help' for more information.\n", program_name);
		return;
	}

	info("Usage: %s [-h] [-r] [[-g] <controllers>:<path>] ...\n", program_name);
	info("Remove control group(s)\n");
	info("  -g <controllers>:<path>	Control group to be removed (-g is optional)\n");
	info("  -h, --help			Display this help\n");
	info("  -r, --recursive		Recursively remove all subgroups\n");
#ifdef WITH_SYSTEMD
	info("  -b				Ignore default systemd delegate hierarchy\n");
#endif
}

/*
 * Skip adding controller which points to the same cgroup when delete
 * cgroup with specifying multi controllers. Just skip controller which
 * cgroup and hierarchy number is same
 */
static int skip_add_controller(int counter, int *skip, struct ext_cgroup_record *ecg_list)
{
	struct controller_data info;
	void *handle;
	int ret = 0;
	int k;

	/* find out hierarchy number of added cgroup */
	ecg_list[counter].h_number = 0;

	ret = cgroup_get_all_controller_begin(&handle, &info);
	while (ret == 0) {
		if (!strcmp(info.name, ecg_list[counter].name)) {
			/* hierarchy number found out, set it */
			ecg_list[counter].h_number = info.hierarchy;
			break;
		}
		ret = cgroup_get_all_controller_next(&handle, &info);
	}
	cgroup_get_all_controller_end(&handle);

	/* deal with cgroup_get_controller_begin/next ret values */
	if (ret == ECGEOF)
		ret = 0;
	if (ret) {
		err("cgroup_get_controller_begin/next failed(%s)\n", cgroup_strerror(ret));
		return ret;
	}

	/* found out whether the hierarchy should be skipped */
	*skip = 0;
	for (k = 0; k < counter; k++) {
		if ((!strcmp(ecg_list[k].name, ecg_list[counter].name)) &&
		    (ecg_list[k].h_number == ecg_list[counter].h_number)) {
			/* we found a control group in the same hierarchy */
			if (strcmp(ecg_list[k].controller, ecg_list[counter].controller)) {
				/*
				 * it is a different controller -> if there
				 * is not one cgroup for the same controller,
				 * skip it
				 */
				*skip = 1;
			} else {
				/*
				 * there is the identical group,controller
				 * pair don't skip it
				 */
				*skip = 0;
				return ret;
			}
		}
	}

	return ret;
}


int main(int argc, char *argv[])
{
	int ignore_default_systemd_delegate_slice = 0;
	struct cgroup_group_spec **cgroup_list = NULL;
	struct ext_cgroup_record *ecg_list = NULL;
	struct cgroup_controller *cgc;
	struct cgroup *cgroup;

	int final_ret = 0;
	int counter = 0;
	int flags = 0;
	int max = 0;
	int ret = 0;
	int i, j, c;
	int skip;

	if (argc < 2) {
		usage(1, argv[0]);
		exit(EXIT_BADARGS);
	}

	/* initialize libcg */
	ret = cgroup_init();
	if (ret) {
		err("%s: libcgroup initialization failed: %s\n", argv[0], cgroup_strerror(ret));
		goto err;
	}

	cgroup_list = calloc(argc, sizeof(struct cgroup_group_spec *));
	if (cgroup_list == NULL) {
		err("%s: out of memory\n", argv[0]);
		ret = -1;
		goto err;
	}

	ecg_list = calloc(argc, sizeof(struct ext_cgroup_record));
	if (ecg_list == NULL) {
		err("%s: out of memory\n", argv[0]);
		ret = -1;
		goto err;
	}

	/* Parse arguments */
#ifdef WITH_SYSTEMD
	while ((c = getopt_long(argc, argv, "rhg:b", long_options, NULL)) > 0) {
		switch (c) {
		case 'b':
			ignore_default_systemd_delegate_slice = 1;
			break;
#else
	while ((c = getopt_long(argc, argv, "rhg:", long_options, NULL)) > 0) {
		switch (c) {
#endif
		case 'r':
			flags |= CGFLAG_DELETE_RECURSIVE;
			break;
		case 'g':
			ret = parse_cgroup_spec(cgroup_list, optarg, argc);
			if (ret != 0) {
				err("%s: error parsing cgroup '%s'", argv[0], optarg);
				ret = EXIT_BADARGS;
				goto err;
			}
			break;
		case 'h':
			usage(0, argv[0]);
			ret = 0;
			goto err;
		default:
			usage(1, argv[0]);
			ret = EXIT_BADARGS;
			goto err;
		}
	}

	/* this is false always for disable-systemd */
	if (!ignore_default_systemd_delegate_slice)
		cgroup_set_default_systemd_cgroup();

	/* parse groups on command line */
	for (i = optind; i < argc; i++) {
		ret = parse_cgroup_spec(cgroup_list, argv[i], argc);
		if (ret != 0) {
			err("%s: error parsing cgroup '%s'\n", argv[0], argv[i]);
			ret = EXIT_BADARGS;
			goto err;
		}
	}

	/* for each cgroup to be deleted */
	for (i = 0; i < argc; i++) {
		if (!cgroup_list[i])
			break;

		/* create the new cgroup structure */
		cgroup = cgroup_new_cgroup(cgroup_list[i]->path);
		if (!cgroup) {
			ret = ECGFAIL;
			err("%s: can't create new cgroup: %s\n", argv[0], cgroup_strerror(ret));
			goto err;
		}

		/* add controllers to the cgroup */
		j = 0;
		while (cgroup_list[i]->controllers[j]) {
			skip = 0;
			/*
			 * save controller name, cg name and hierarchy
			 * number to determine whether we should skip
			 * adding controller
			 */
			if (counter == max) {
				/*
				 * there is not enough space to store
				 * them, create it
				 */
				max = max + argc;
				ecg_list = (struct ext_cgroup_record *)
					realloc(ecg_list, max * sizeof(struct ext_cgroup_record));
				if (!ecg_list) {
					err("%s: not enough memory\n", argv[0]);
					final_ret = -1;
					goto err;
				}
			}

			strncpy(ecg_list[counter].controller,
				cgroup_list[i]->controllers[j], FILENAME_MAX);
			ecg_list[counter].controller[FILENAME_MAX - 1] = '\0';

			strncpy(ecg_list[counter].name,
				cgroup_list[i]->path, FILENAME_MAX);
			ecg_list[counter].name[FILENAME_MAX - 1] = '\0';

			ret = skip_add_controller(counter, &skip, ecg_list);
			if (ret)
				goto err;

			if (skip) {
				/* don't add the controller, goto next one */
				goto next;
			}

			cgc = cgroup_add_controller(cgroup, cgroup_list[i]->controllers[j]);
			if (!cgc) {
				ret = ECGFAIL;
				err("%s: controller %s can't be added\n", argv[0],
				    cgroup_list[i]->controllers[j]);
				cgroup_free(&cgroup);
				goto err;
			}
next:
			counter++;
			j++;
		}

		ret = cgroup_delete_cgroup_ext(cgroup, flags);
		/* Remember the errors and continue, try to remove all groups. */
		if (ret != 0) {
			err("%s: cannot remove group '%s': %s\n", argv[0], cgroup->name,
			    cgroup_strerror(ret));
			final_ret = ret;
		}
		cgroup_free(&cgroup);
	}

	ret = final_ret;

err:
	if (ecg_list)
		free(ecg_list);

	if (cgroup_list) {
		for (i = 0; i < argc; i++) {
			if (cgroup_list[i])
				cgroup_free_group_spec(cgroup_list[i]);
		}
		free(cgroup_list);
	}

	return ret;
}
